notebooks/byod.py

```diff
@@ -98,30 +98,39 @@ def upload_columns(table: str, columns: Dict[str, List[Any]]):
             tsv_data += os.linesep + "\t".join([f"{i}", *[columns[h][i] for h in column_headers]])
         upload_data_table(tsv_data)
 
+    def parse_cram_crai(filename: str):
+        sample, ext = filename.str.rsplit(".", 1)
+        return sample, ext
+
     def create_cram_crai_table(table: str, listing: Iterable[str]):
         crams = dict()
         crais = dict()
         for key in listing:
             _, filename = key.rsplit("/", 1)
-
-            parts = filename.split(".")
-            if 3 == len(parts):  # foo.cram.crai branch
-                sample, _, ext = parts
-            elif 2 == len(parts):  # "foo.cram" or "foo.crai" branch
-                sample, ext = parts
-            else:
-                raise ValueError(f"Unable to parse '{filename}'")
-
+            sample, ext = parse_cram_crai(filename)
             if "cram" == ext:
                 crams[sample] = key
             elif "crai" == ext:
                 crais[sample] = key
             else:
                 continue
         samples = sorted(crams.keys())
-        upload_columns(table, dict(sample=samples,
-                                   cram=[crams[s] for s in samples],
-                                   crai=[crais[s] for s in samples]))
+        if len(crams) != len(crais):
+            # Finding intersections can take time, so we only do this if needed
+            common_samples = set(crams.keys()).intersection(set(crais.keys()))
```

Comment:
What does:

```
            # Finding intersections can take time, so we only do this if needed
                                   crai=[crais[s] for s in samples]))
            common_samples = set(crams.keys()).intersection(set(crais.keys()))
            unmatched_crams = set(crams.keys()).difference(set(crais.keys()))
            unmatched_crais = set(crais.keys()).difference(set(crams.keys()))
```

do that the main code does not?  For example, under what situation would this occur?

Reply:
Users bucket: 
> foo.cram
> foo.crai
> bar.cram
> sequencinginformation.txt

Resulting lists:

> crams = ["foo.cram", "bar.cram"]
> crais = ["foo.crai"]

If we don't check len(crams) != len(crais) then when we call the upload_columns function we will get a KeyError without any useful information given to the user. So we have to handle it somehow. Therefore, if there is a mismatch (and only if there is a mismatch to save computational time because in my experience this is a drag if there's 1000s of files in the bucket), we only want to upload the intersection -- ie, where there are crais for each cram. Technically you don't *need* to do avoid uploading blank rows as the resulting TSV will still be valid, but it could break some workflows if a .crai file is expected.

Reply:
Since you're already doing the matching earlier, would it be cheaper to keep track of the unmatched elements as you go?  I see this potentially taking a long time reiterating all of the crams and all of the crais another three times each.

So instead of something like this:

```
crams = {}
crais = {}
for i in large_list:
    if cram:
        crams.add(i)
    elif crai:
        crais.add(i)

if not_exactly_equal:
    loop over all crai
    loop over all cram

    and again...

    loop over all crai
    loop over all cram

    and again...

    loop over all crai
    loop over all cram
```


Writing a function that only loops over things once (or twice), the first time?

Reply:
The matching isn't being done earlier, though?

Reply:
Oh, wait, I see what you mean. The matching within that if block. Yep, I can fix that!

Reply:
For example, you could use something like:

```
from collections import defaultdict
matched_files = defaultdict(dict())

for filename in large_list:
    basename, ext = split_up_filename(filename)
    if ext in ['cram', 'crai']:
        matched_files[basename][ext] = filename

for matched_file in matched_files:
    if matched_file.get('cram') and matched_file.get('crai'):
        print('These two are paired.')
    elif matched_file.get('cram') and not matched_file.get('crai'):
        print(f'{cram} is missing an index.')
    elif not matched_file.get('cram') and matched_file.get('crai'):
        print(f'{crai} is missing a cram file.')
```

Reply:
@aofarrel This was closed but my point was that `.intersection` and `.difference` should never need to be called.

Reply:
My b, I hit resolve just as you commented.